[#17 #18] 사용자 로그인, 로그아웃, 회원가입 기능을 구현한다.

rlarltj opened the pull request:
### 🍀 목적
- JWT 설정 및 적용

### 🌹 추가사항
- Jwt 의존성 추가: io.jsonwebtoken
https://www.libhunt.com/compare-jwt-java-vs-jjwt?ref=compare
- JwtAuthenticationFilter 추가
- TokenProvider 추가
- AuthorityNotFoundException 추가

### ☕ 변경사항
- AuthController로 기능 분리

### ❓ pr 포인트
- JWT 세팅 어떤가요

rlarltj committed:
feat: Authority, UserAuthority 엔티티 추가

rlarltj committed:
refactor: AuthController로 기능 분리

rlarltj committed:
feat: JWT 설정 및 유틸 구현
- TokenProvider
- JwtAuthenticationFilter
- TokenResponse(DTO)

devYSK commented on build.gradle:
```diff
@@ -84,6 +84,10 @@ dependencies {
     runtimeOnly 'com.h2database:h2'
 
+    // jwt
+    implementation group: 'io.jsonwebtoken', name: 'jjwt-api', version: '0.11.5'
```
LGTM

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,52 @@
+		filterChain.doFilter(servletRequest, servletResponse);
+	}
+
+	private String resolveToken(HttpServletRequest request) {
+		String bearerToken = request.getHeader("Authorization");
+
+		if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
+			return bearerToken.substring(7);
+		}
+
+		return null;
+	}
```
null을 반환하는것보다 Optional 또는 400이나 401 응답은 어떨까요? 

null 채로 반환하는건 좋지 않은것 같습니다 때문에 밑에 로직에서 StringUtils.hasText()라는 jwt 토큰과는 관련없는 문자열 검사를 하고 있네요.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,105 @@
+	private final Key key;
+	private final long tokenValidityInSeconds;
+
+	public TokenProvider(@Value("${jwt.secret}") String secretKey,
+		@Value("${token-validity-in-seconds}") long tokenValidityInSeconds) {
+
```
Properties 클래스로 관리할 수 있을 것 같습니다.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,105 @@
+@Component
+@Slf4j
+public class TokenProvider {
+	private static final String AUTHORITIES_KEY = "auth";
+	private final Key key;
+	private final long tokenValidityInSeconds;
```
key는 code보다는 property 파일 또는 다른 방식으로  관리할 수 있을것 같습니다.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,105 @@
+
+		Date validity = new Date(new Date().getTime() + tokenValidityInSeconds);
+
+		return Jwts.builder()
+			.setSubject(authentication.getName())
+			.claim(AUTHORITIES_KEY, authorities)
+			.signWith(key, SignatureAlgorithm.HS512)
```
claim에 userId나 email같은 정보를 넣어둔다면 추후 다른 로직에서 쉽게 이용할 수 있을것 같습니다. 
보안에 민감하지 않은 정보들로요! 어떻게 생각하시나요?

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	public String generateToken(Authentication authentication) {
+		String authorities = authentication.getAuthorities().stream()
+			.map(GrantedAuthority::getAuthority)
+			.collect(Collectors.joining(","));
```
권한은 보통 []로 관리하는것이 좋아보입니다.

클라이언트에서 권한들을 꺼낼 때 array가 아닌 ,로 join된 String이라면 한번 더 ,로 분할해야 하는 일이 생길 것 같아요.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,105 @@
+			Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(token);
+			return true;
+
+		} catch (SecurityException | MalformedJwtException e) {
+			log.info("잘못된 JWT 서명입니다.");
+		} catch (ExpiredJwtException e) {
+			log.info("만료된 JWT 토큰입니다.");
+		} catch (UnsupportedJwtException e) {
+			log.info("지원되지 않는 JWT 토큰입니다.");
+		} catch (IllegalArgumentException e) {
+			log.info("JWT 토큰이 빈 값입니다.");
+		}
```
Log를 찍고  Jwt or Authentication 예외를 감싸 던져서 처리할 수 있을것같아요.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,105 @@
+
+			throw new AuthorityNotFoundException("클레임에 권한정보가 존재하지 않습니다..");
+
+		} else {
+			Collection<? extends GrantedAuthority> authorities =
+				Arrays.stream(claims.get(AUTHORITIES_KEY).toString().split(","))
+					.map(SimpleGrantedAuthority::new)
+					.collect(Collectors.toList());
+
+			User principal = new User(claims.getSubject(), "", authorities);
+			return new UsernamePasswordAuthenticationToken(principal, accessToken, authorities);
+		}
```
UserPrincipal을 상속받거나 UserDetails를 구현하여 우리의 User를 담는 Adapter 객체를 만들어 보는것은 어떨까요? 

UsernamePasswordAuthentiocationToken도 우리가 Authentication을 구현해서 다르게 사용할 수 있을것 같아요.

tinajeong replied:
영향 받는 코드들이 있어 다음 PR에 적용하겠습니다!

devYSK reviewed (changes requested):
안녕하세요 찬미님 기서님 고생 많으셨습니다! 

적어놓은 리뷰에 대해 고민하시면 더 좋은 코드가 나올 것 같습니다.

rlarltj committed:
refactor: jwt 설정 관련
- configuration-processor 의존성 추가
- JwtConfigure 클래스로 속성 값 관리
- resolveToken() Optional 반환으로 변경
- List로 Authorities 관리
- Claim 객체에 사용자 Email 정보 추가
- JwtValidateException 커스텀 예외 추가

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,118 @@
+			.parseClaimsJws(accessToken)
+			.getBody();
+
+		if (claims.get(AUTHORITIES_KEY) == null) {
+
+			throw new AuthorityNotFoundException("클레임에 권한정보가 존재하지 않습니다.");
+
```
필요없이 \n이 많은것같아요!

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,118 @@
+				roles.stream().map(SimpleGrantedAuthority::new)
+					.collect(Collectors.toList());
+
+			principal = new org.springframework.security.core.userdetails.User(claims.getSubject(), "", authorities);
```
todo : UserDetails 인터페이스를 구현한 우리의 어뎁터 객체 로 new 하면 좋을것같네요.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,118 @@
+
+	private final JwtConfigure jwtConfigure;
+
+	private org.springframework.security.core.userdetails.User principal;
```
이친구가 field로 존재해야하만 하는 이유가 궁금합니다!

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,118 @@
+		byte[] keyBytes = Decoders.BASE64.decode(jwtConfigure.getSecret());
+
+		this.key = Keys.hmacShaKeyFor(keyBytes);
+		this.tokenValidityInSeconds = jwtConfigure.getTokenValidityInSeconds() * 1000;
```
1000을 파라미터로 받아도 좋을것같네요 jwtConfigure가 연산하는 책임을 가지도록.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/TokenProvider.java:
```diff
@@ -0,0 +1,118 @@
+@Component
+public class TokenProvider {
+
+	private static final String AUTHORITIES_KEY = "auth";
```
AUTHORITIES_KEY는 중요한  친구같은데 code에 노출되어도 될까요..?

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/JwtConfigure.java:
```diff
@@ -0,0 +1,22 @@
+
+@Configuration
+@RequiredArgsConstructor
+@Getter
+@ConfigurationProperties(prefix = "jwt")
+@ConstructorBinding
+public class JwtConfigure {
+
+	private final String header;
```
LGTM

but 

생성자 관련된 @RequiredArgsConstructor 가 아래로 내려가야할거같아요

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,55 @@
+		ServletException {
+		HttpServletRequest httpServletRequest = (HttpServletRequest)servletRequest;
+		String jwt = resolveToken(httpServletRequest);
+		String requestURI = httpServletRequest.getRequestURI();
```
이 requestURI는 else {} 구문에서만 쓰인다면 저 안으로 들어가도 될것같아요.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,55 @@
+		String jwt = resolveToken(httpServletRequest);
+		String requestURI = httpServletRequest.getRequestURI();
+
+		if (StringUtils.hasText(jwt) && tokenProvider.validateToken(jwt)) {
```
validateToken 또는 resolveToken 안에  hasText()가 들어가면 좋을거같아요.

resolveToken에서 예외가 발생하지 않았다는건 Bearer로 시작하는 문자열 토큰이 존재한다는 이야기이고. 그렇다면 hasText에는 딱히 걸릴일이 없으므로 if문이 간단해질것같네요.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/entity/Authority.java:
```diff
@@ -0,0 +1,25 @@
+@AllArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class Authority extends AbstractTimeColumn {
+
+	@Id
+	@Column(name = "authority_name", length = 10)
+	private String authorityName;
+}
```
정의한 enum(role?)로 사용해도 될듯합니다.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/dto/TokenResponse.java:
```diff
@@ -0,0 +1,13 @@
+@NoArgsConstructor
+public class TokenResponse {
+
+	private String token;
```
final

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/controller/UserController.java:
```diff
@@ -23,54 +23,21 @@
-	}
-
-	@GetMapping(value = {"/users/me"})
+	@GetMapping(value = {"/me"})
```
`@AuthenticationPrincial`을 도입하면 어떨까요? 

필터에서 인증이 완료되면 UsernamePasswordAuthenticationToken 이라는 authentication 객체가 생성되고, 그 안에 우리의 User 도메인이나 어뎁터가 있어도 될것같네요.

간단한 정보는 도메인을통해 얻을 수 있을것 같아요.

devYSK reviewed (changes requested):
고생하셨습니다! 수정하면 좋겠는 부분이 몇가지 있어요

rlarltj committed:
refactor: jwt 설정 변경
- authoritiesKey 프로퍼티 클래스로 관리
- Authority 엔티티 타입을 Enum으로 변경

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/jwt/JwtConfigure.java:
```diff
@@ -4,19 +4,39 @@
 
 	private final long tokenValidityInSeconds;
+
+	public String getAuthoritiesKey() {
+		return authoritiesKey;
+	}
+
+	public String getHeader() {
+		return header;
+	}
+
+	public String getSecret() {
+		return secret;
+	}
+
```
`@Getter`는 필드마다 붙일 수도 있습니다!

devYSK committed:
Merge branch 'develop' into feature/user-security
